core/src/main/java/org/infinispan/util/concurrent/locks/impl/DefaultPendingLockManager.java

```diff
@@ -0,0 +1,387 @@
+                                                            long expectedEndTime) throws InterruptedException {
+      if (transactionsToCheck.isEmpty()) {
+         return null;
+      }
+      for (PendingTransaction tx : transactionsToCheck) {
+         long remaining;
+         if ((remaining = timeService.remainingTime(expectedEndTime, TimeUnit.MILLISECONDS)) > 0) {
+            if (!tx.cacheTransaction.waitForLockRelease(remaining)) {
+               return tx;
+            }
+         }
+      }
+      return null;
+   }
+
+   private Collection<PendingTransaction> getTransactionWithLockedKey(int transactionTopologyId,
+                                                                      Object key,
+                                                                      GlobalTransaction globalTransaction) {
+
+      Function<CacheTransaction, PendingTransaction> mapper = transaction -> {
+         if (transaction.getTopologyId() < transactionTopologyId &&
+               !transaction.getGlobalTransaction().equals(globalTransaction) &&
+               transaction.containsLockOrBackupLock(key)) {
+            return new PendingTransaction(transaction, key);
+         }
+         return null;
+      };
+
+      return filterAndCollectTransactions(mapper);
+   }
+
+   private Collection<PendingTransaction> getTransactionWithAnyLockedKey(int transactionTopologyId,
+                                                                         Collection<Object> keys,
+                                                                         GlobalTransaction globalTransaction) {
+
+      Function<CacheTransaction, PendingTransaction> filter = transaction -> {
+         if (transaction.getTopologyId() < transactionTopologyId &&
+               !transaction.getGlobalTransaction().equals(globalTransaction)) {
+            Object key = transaction.findAnyLockedOrBackupLocked(keys.stream());
+            return key == null ? null : new PendingTransaction(transaction, key);
+         }
+         return null;
+      };
+
+      return filterAndCollectTransactions(filter);
+   }
+
+   private Collection<PendingTransaction> filterAndCollectTransactions(Function<CacheTransaction, PendingTransaction> mapper) {
+      final Collection<? extends CacheTransaction> localTransactions = transactionTable.getLocalTransactions();
+      final Collection<? extends CacheTransaction> remoteTransactions = transactionTable.getRemoteTransactions();
+      final int totalSize = localTransactions.size() + remoteTransactions.size();
+      if (totalSize == 0) {
+         return Collections.emptyList();
+      }
+      final List<PendingTransaction> allTransactions = new ArrayList<>(totalSize);
+      final Collector<PendingTransaction, ?, Collection<PendingTransaction>> collector = Collectors.toCollection(() -> allTransactions);
+
+
+      if (!localTransactions.isEmpty()) {
+         localTransactions.stream().map(mapper).filter(tx -> tx != null).collect(collector);
```

Comment:
Mixing mapping and filtering looks kind of fishy to me, better use `forEach` I think.